### PR TITLE
[2443] Use dynamic handles

### LINKED ISSUE #2443: Make handle appear only conditionally
By conditionally, I mean something like on selection or on hover, I don't know, it will have to be defined

### PR DESCRIPTION
**Pull request by mcharfadi**
Bug: https://github.com/eclipse-sirius/sirius-web/issues/2443

# Pull request template

## General purpose
What is the main goal of this pull request?
- [ ] Bug fixes
- [ ] New features
- [ ] Documentation
- [ ] Cleanup
- [ ] Tests
- [ ] Build / releng

## Project management
- [ ] Has the pull request been added to the relevant project and milestone? (Only if you know that your work is part of a specific iteration such as the current one)
- [ ] Have the `priority:` and `pr:` labels been added to the pull request? (In case of doubt, start with the labels `priority: low` and `pr: to review later`)
- [ ] Have the relevant issues been added to the pull request?
- [ ] Have the relevant labels been added to the issues? (`area:`, `difficulty:`, `type:`)
- [ ] Have the relevant issues been added to the same project and milestone as the pull request?
- [ ] Has the `CHANGELOG.adoc` been updated to reference the relevant issues?
- [ ] Have the relevant API breaks been described in the `CHANGELOG.adoc`? (Including changes in the GraphQL API)
- [ ] In case of a change with a visual impact, are there any screenshots in the `CHANGELOG.adoc`? For example in `doc/screenshots/2022.5.0-my-new-feature.png`

## Architectural decision records (ADR)
- [ ] Does the title of the commit contributing the ADR start with `[doc]`?
- [ ] Are the ADRs mentioned in the relevant section of the  `CHANGELOG.adoc`?

## Dependencies
- [ ] Are the new / upgraded dependencies mentioned in the relevant section of the `CHANGELOG.adoc`?
- [ ] Are the new dependencies justified in the `CHANGELOG.adoc`?

## Frontend

This section is not relevant if your contribution does not come with changes to the frontend.

### General purpose
- [ ] Is the code properly tested? (Plain old JavaScript tests for business code and tests based on React Testing Library for the components)

### Typing
We need to improve the typing of our code, as such, we require every contribution to come with proper TypeScript typing for both changes contributing new files and those modifying existing files.
Please ensure that the following statements are true for each file created or modified (this may require you to improve code outside of your contribution).

- [ ] Variables have a proper type
- [ ] Functions’ arguments have a proper type
- [ ] Functions’ return type are specified
- Hooks are properly typed:
	- [ ] `useMutation<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useQuery<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useSubscription<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useMachine<CONTEXT_TYPE, EVENTS_TYPE>(…)`
	- [ ] `useState<STATE_TYPE>(…)`
- [ ] All components have a proper typing for their props
- [ ] No useless optional chaining with `?.` (if the GraphQL API specifies that a field cannot be `null`, do not treat it has potentially `null` for example)
- [ ] Nullable values have a proper type (for example `let diagram: Diagram | null = null;`)

## Backend

This section is not relevant if your contribution does not come with changes to the backend.

### General purpose
- [ ] Are all the event handlers tested?
- [ ] Are the event processor tested?
- [ ] Is the business code (services) tested?
- [ ] Are diagram layout changes tested?

### Architecture
- [ ] Are data structure classes properly separated from behavioral classes?
- [ ] Are all the relevant fields final?
- [ ] Is any data structure mutable? If so, please write a comment indicating why
- [ ] Are behavioral classes either stateless or side effect free?

### How to test this PR?
_Please describe here the various use cases to test this pull request_

- [ ] Has the Kiwi TCMS test suite been updated with tests for this contribution?

### CODE REVIEW

File: CHANGELOG.adoc [hunk 1]
```diff
@@ -57,6 +57,7 @@ This coupling will be removed in the near future as a consequence, the dependenc
 - https://github.com/eclipse-sirius/sirius-web/issues/2410[#2410] [diagram] Do a fit to screen on a diagram opening
 - https://github.com/eclipse-sirius/sirius-web/issues/2268[#2268] [diagram] Create child nodes where the user has clicked.
 - https://github.com/eclipse-sirius/sirius-web/issues/2475[#2475] [diagram] Prevent nodes to move each diagram refresh event. Only the size is still updated
+- https://github.com/eclipse-sirius/sirius-web/issues/2443[#2443] [diagram] Use dynamic handles.
```
Comment (pcdavid) [inline CHANGELOG.adoc]: A little short for such a visible change. I'm not sure anyone outside of the dev team can know what this is about.

File: packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/node/RectangularNode.tsx [hunk 2]
```diff
@@ -40,25 +46,35 @@ const rectangularNodeStyle = (
   if (selected) {
     rectangularNodeStyle.outline = `${theme.palette.primary.main} solid 1px`;
   }
+
+  if (hoveredNode && !selected) {
+    rectangularNodeStyle.outline = `${theme.palette.primary.main} dashed 0.5px`;
+  }
+
```
Comment (pcdavid) [inline packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/node/RectangularNode.tsx]: Not sure how this is related to connection handles? And why only for Rectangular node?

File: packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/node/RectangularNode.tsx [hunk 3]
```diff
@@ -40,25 +46,35 @@ const rectangularNodeStyle = (
...
+  const { newConnectionStyleProvider } = useConnector();
+  const { hoveredNode } = useContext<NodeContextValue>(NodeContext);
 
   const handleOnDrop = (event: React.DragEvent) => {
     onDrop(event, id);
   };
 
+  const isHoveredNode = () => {
+    return !!hoveredNode && hoveredNode.id === id;
+  };
```
Comment (pcdavid) [inline packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/node/RectangularNode.tsx]: Any reason to define this as a function instead of simply like this?

```java
const isHoveredNode = hoveredNode?.id === id;
```

File: packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/node/RectangularNode.tsx [hunk 4]
```diff
@@ -67,38 +83,9 @@ export const RectangularNode = memo(({ data, isConnectable, id, selected }: Node
...
-          id={`handle--${id}--bottom`}
-          type="source"
-          position={Position.Bottom}
-          isConnectable={isConnectable}
-          style={newConnectionStyleProvider.getHandleStyle(data.descriptionId)}
-          onMouseDown={onConnectionStartElementClick}
-        />
+        {selected ? <ConnectionSourceHandles nodeId={id}></ConnectionSourceHandles> : null}
+        <ConnectionTargetHandles nodeId={id}></ConnectionTargetHandles>
+        <ConnectionHandles nodeId={id} customHandles={data.customHandles}></ConnectionHandles>
```
Comment (pcdavid) [inline packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/node/RectangularNode.tsx]: Simpler syntax for empty elements:

```
{selected ? <ConnectionSourceHandles nodeId={id} /> : null}
<ConnectionTargetHandles nodeId={id} />
<ConnectionHandles nodeId={id} customHandles={data.customHandles} />
```

File: packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/node/ListNode.tsx [hunk 5]
```diff
@@ -70,38 +73,9 @@ export const ListNode = memo(({ data, isConnectable, id, selected }: NodeProps<L
...
-          id={`handle--${id}--bottom`}
-          type="source"
-          position={Position.Bottom}
-          isConnectable={isConnectable}
-          style={newConnectionStyleProvider.getHandleStyle(data.descriptionId)}
-          onMouseDown={onConnectionStartElementClick}
-        />
+        {selected ? <ConnectionSourceHandles nodeId={id}></ConnectionSourceHandles> : null}
+        <ConnectionTargetHandles nodeId={id}></ConnectionTargetHandles>
+        <ConnectionHandles nodeId={id} customHandles={data.customHandles}></ConnectionHandles>
```
Comment (pcdavid) [inline packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/node/ListNode.tsx]: Simpler syntax for empty elements:

```
{selected ? <ConnectionSourceHandles nodeId={id} /> : null}
<ConnectionTargetHandles nodeId={id} />
<ConnectionHandles nodeId={id} customHandles={data.customHandles} />
```

File: packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/node/ImageNode.tsx [hunk 6]
```diff
@@ -59,38 +62,9 @@ export const ImageNode = memo(({ data, isConnectable, id, selected }: NodeProps<
...
-        id={`handle--${id}--bottom`}
-        type="source"
-        position={Position.Bottom}
-        isConnectable={isConnectable}
-        style={newConnectionStyleProvider.getHandleStyle(data.descriptionId)}
-        onMouseDown={onConnectionStartElementClick}
-      />
+      {selected ? <ConnectionSourceHandles nodeId={id}></ConnectionSourceHandles> : null}
+      <ConnectionTargetHandles nodeId={id}></ConnectionTargetHandles>
+      <ConnectionHandles nodeId={id} customHandles={data.customHandles}></ConnectionHandles>
```
Comment (pcdavid) [inline packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/node/ImageNode.tsx]: Simpler syntax for empty elements:

```
{selected ? <ConnectionSourceHandles nodeId={id} /> : null}
<ConnectionTargetHandles nodeId={id} />
<ConnectionHandles nodeId={id} customHandles={data.customHandles} />
```

Comment (pcdavid) [reply packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/node/ImageNode.tsx]: These 3 elements (ConnectionSourceHandles, ConnectionTargetHandles and ConnectionHandles) are always used together and in the same exact way. Couldn't this be a single React component?

File: packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/node/CustomHandle.tsx [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+        position: 'relative',
+        transform: 'none',
+        left: 'auto',
+        opacity: '0',
+        pointerEvents: 'none',
+      };
+  }
+};
+
+export const CustomHandle = ({ nodeId, id, type, position, handleIndex, ...props }: CustomHandleProps) => {
```
Comment (pcdavid) [inline packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/node/CustomHandle.tsx]: We normally use the terminology "CustomXXX" for concepts that applications can extend by contributing their own (custom) versions. For example custom _widgets_, custom _menu entries_, custom _node types_, etc.
This should probably be renamed to better reflect the component's purpose.

File: packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/node/CustomHandle.tsx [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+        pointerEvents: 'none',
+      };
+    case Position.Bottom:
+      return {
+        position: 'relative',
+        transform: 'none',
+        left: 'auto',
+        opacity: '0',
+        pointerEvents: 'none',
+      };
```
Comment (pcdavid) [inline packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/node/CustomHandle.tsx]: They're all *almost* identical, but the way they are defined it's difficult to see the actual differences.
Proposal:
```js
  const style: React.CSSProperties = {
    position: 'relative',
    transform: 'none',
    opacity: '0',
    pointerEvents: 'none',
  };
  switch (position) {
    case Position.Left:
    case Position.Right:
      style.top = 'auto';
      break;
    case Position.Top:
    case Position.Bottom:
      style.left = 'auto';
      break;
  }
  return style;
```

File: packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/node/CustomHandle.tsx (hunk 7 above)
Comment (pcdavid) [inline packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/node/CustomHandle.tsx]: `handleIndex` is not used; it can be removed from `CustomHandleProps`.

File: packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/node/CustomHandle.types.ts [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *     Obeo - initial API and implementation
+ *******************************************************************************/
+import { HandleProps, HandleType, Position } from 'reactflow';
+export interface CustomHandleProps extends HandleProps {
+  id: string;
+  nodeId: string;
+  handleIndex: number;
```
Comment (pcdavid) [inline packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/node/CustomHandle.types.ts]: `handleIndex` does not seem as needed.

File: packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/node/NodeContext.tsx [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+ *
+ * Contributors:
+ *     Obeo - initial API and implementation
+ *******************************************************************************/
+
+import React, { useState } from 'react';
+import { Node } from 'reactflow';
+import { NodeData } from '../DiagramRenderer.types';
+import { NodeContextProviderProps, NodeContextProviderState, NodeContextValue } from './NodeContext.types';
+const defaultValue: NodeContextValue = {
```
Comment (pcdavid) [inline packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/node/NodeContext.tsx:18]: This whole `NodeContext` seem only to be used to set a dashed outline on reactangular nodes on hover.
* I'm not sure how this is related to the subject of connection handles.
* Why only on Rectangular nodes?

File: packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/node/ConnectionSourceHandles.tsx [hunk 11]
```diff
@@ -0,0 +1,260 @@
...
+            }
+          }
+        }
+      }
+    }
+  }
+`;
+
+const customConnectionHandlesStyle = (position: Position, theme: Theme): React.CSSProperties => {
+  switch (position) {
```
Comment (pcdavid) [inline packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/node/ConnectionSourceHandles.tsx]: Difficult to read what's the same (most properties) and what is different between the 4 cases.

File: packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/node/ConnectionSourceHandles.tsx [hunk 12]
```diff
@@ -0,0 +1,260 @@
...
+      return {
+        position: 'absolute',
+        borderRadius: '0',
+        right: '-15px',
+        border: 'solid black',
+        borderWidth: '0 2px 2px 0',
+        display: 'inline-block',
+        padding: '2px',
+        transform: 'rotate(-45deg)',
+        WebkitTransform: 'rotate(-45deg)',
```
Comment (pcdavid) [inline packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/node/ConnectionSourceHandles.tsx]: Is the Webkit-specific property really needed? I don't see any other uses elsewhere in our code. On the contrary, we have several places where we use the standard `transform:` and do not add a Webkit-specific variant.

File: packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/node/ConnectionSourceHandles.tsx [hunk 13]
```diff
@@ -0,0 +1,260 @@
...
+
+      if (toolSectionsError) {
+        addErrorMessage(toolSectionsError.message);
+      }
+    }
+  }, [toolSectionsLoading, toolSectionsData, toolSectionsError]);
+
+  return (
+    <>
+      {candidates.length > 0 ? (
```
Comment (pcdavid) [inline packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/node/ConnectionSourceHandles.tsx]: Why 4 separate tests for the same condition?

```jsx
  if (candidates.length > 0) {
    return (
      <>
        <Handle />
        <Handle />
        <Handle />
        <Handle />
      </>
    );
  } else {
    return null;
  }
```
This applies to a few other places.

File: packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/converter/convertDiagram.ts [hunk 14]
```diff
@@ -122,7 +144,7 @@ const toIconLabelNode = (
   const style = gqlNode.style as GQLIconLabelNodeStyle;
   const { id, label } = gqlNode;
   const labelStyle = label.style;
-
+  const customHandles: CustomHandleProps[] = [];
```
Comment (pcdavid) [inline packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/converter/convertDiagram.ts]: The name `customHandle` (here and elsewhere) should probably be changed. It does not tell what these handles are/do.

File: packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/node/ConnectionTargetHandles.tsx [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+    pointerEvents: 'all',
+    cursor: 'crosshair',
+  };
+};
+const connectionNodeIdSelector = (state) => state.connectionNodeId;
+export const ConnectionTargetHandles = ({ nodeId }: ConnectionTargetHandlesProps) => {
+  const connectionNodeId = useStore(connectionNodeIdSelector);
+  const isConnecting = !!connectionNodeId;
+  return (
+    <>
```
Comment (pcdavid) [inline packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/node/ConnectionTargetHandles.tsx]: Why a fragment? There is at most one component.

File: packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/node/ConnectionTargetHandles.tsx [hunk 16]
```diff
@@ -0,0 +1,51 @@
...
+        <Handle
+          id={`handle--${nodeId}--target`}
+          type="target"
+          position={Position.Top}
+          style={targetHandleNodeStyle()}
+          isConnectableEnd={true}
+          isConnectableStart={false}
+        />
+      ) : (
+        ''
```
Comment (pcdavid) [inline packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/node/ConnectionTargetHandles.tsx]: We usually return `null` to render nothing. En empty string will render a text node in the DOM, I'm not sure that's what we want.

File: packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/node/ConnectionHandles.tsx [hunk 17]
```diff
@@ -0,0 +1,138 @@
...
+                nodeId={nodeId}
+                type={customHandle.type}
+                position={customHandle.position}
+                key={customHandle.id}
+                handleIndex={index}
+                isConnectable={true}
+              />
+            );
+          })}
+      </div>
```
Comment (pcdavid) [inline packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/node/ConnectionHandles.tsx]: It's the same pattern repeated 4 times.
It's certainly possible to do something like

```js
for (var position of [Position.Left, Position.Top, Position.Bottom, Position.Right]) {
  <div style={customHandlesStule(position)>...</>
}
```

**Review by pcdavid — CHANGES_REQUESTED** [review]
I have not yet reviewed everything in depth, but I prefer publishing my current remarks now so that we can iterate instead of waiting any longer.

It's too late now, but the PR mixes several changes that should probably have been done separately and iteratively. As it is, it's difficult to follow what code changes correspond to which UI/behavior change between:
* distinguish between the "create connection" handles displayed on a node to start the edge creation process from the "edge source/target handle" displayed *on an edge* to trigger a reconnection;
* only display the "create connection" handles (resp. "edge source/target handle") on a node (resp. edge) when it is selected);
* hightlight the compatible nodes (and fade the incompatible ones) during the edge creation process (actually, this does not seem to work as expected);
* distribute the endpoints of multiple edges which share the same side of a node so that they are visually distinguished.

The names chosen (especially "Custom Handle") do not help.

Cypress tests should also be added.

File: CHANGELOG.adoc [hunk 18]
```diff
@@ -57,6 +57,9 @@ This coupling will be removed in the near future as a consequence, the dependenc
 - https://github.com/eclipse-sirius/sirius-web/issues/2410[#2410] [diagram] Do a fit to screen on a diagram opening
 - https://github.com/eclipse-sirius/sirius-web/issues/2268[#2268] [diagram] Create child nodes where the user has clicked.
 - https://github.com/eclipse-sirius/sirius-web/issues/2475[#2475] [diagram] Prevent nodes to move each diagram refresh event. Only the size is still updated
+- https://github.com/eclipse-sirius/sirius-web/issues/2443[#2443] [diagram] Add dedicated handles to create edges.
+Hide handles between edges and each edges are now placed in differents positions around the node.
+Allow the creation of an edge by droping the connection line on a node.
```
Comment (pcdavid) [inline CHANGELOG.adoc]: Typo: "drop[p]ing"

**Review by pcdavid — CHANGES_REQUESTED** [review]

File: CHANGELOG.adoc [hunk 19]
```diff
@@ -89,6 +89,10 @@ This coupling will be removed in the near future as a consequence, the dependenc
 All icons can now be composed of multiple images, which will be overlayed on top of each other.
 An example of use has been added on the domain entity items.
 - https://github.com/eclipse-sirius/sirius-web/issues/2483[#2483] [diagram] Take the header into account in the React Flow layout.
+- https://github.com/eclipse-sirius/sirius-web/issues/2475[#2475] [diagram] Prevent nodes to move each diagram refresh event. Only the size is still updated
```
Comment (pcdavid) [inline CHANGELOG.adoc]: The line which references #2475 should not be part of this PR. Probably a mistake during a conflict resolution.

File: packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/node/ImageNode.tsx [hunk 20]
```diff
@@ -53,45 +67,17 @@ export const ImageNode = memo(({ data, isConnectable, id, selected }: NodeProps<
...
-        type="source"
-        position={Position.Bottom}
-        isConnectable={isConnectable}
-        style={newConnectionStyleProvider.getHandleStyle(data.descriptionId)}
-        onMouseDown={onConnectionStartElementClick}
-      />
+      {selected ? <ConnectionSourceHandles nodeId={id} /> : null}
+      <ConnectionTargetHandles nodeId={id} />
+      <ConnectionHandles connectionHandles={data.connectionHandles} />
+      {data.label ? <Label diagramElementId={id} label={data.label} faded={data.faded} transform="" /> : null}
```
Comment (pcdavid) [inline packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/node/ImageNode.tsx]: This creates a duplicate label.

File: packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/connector/ConnectorContext.tsx [hunk 21]
```diff
@@ -58,6 +58,7 @@ export const ConnectorContextProvider = ({ children }: ConnectorContextProviderP
       targetCandidates: [],
       isNewConnection: false,
     }));
+    setState((prevState) => ({ ...prevState, connection: null, isNewConnection: false }));
```
Comment (pcdavid) [inline packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/connector/ConnectorContext.tsx]: This seems unnecessary give the `setState` call just above already does the same.

File: packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/node/ImageNode.tsx [hunk 22]
```diff
@@ -40,11 +43,22 @@ const imageNodeStyle = (
...
+
+  const firstUpdate = useRef<boolean>(true);
+
+  useEffect(() => {
+    if (firstUpdate.current && firstUpdate) {
+      firstUpdate.current = false;
+    } else {
+      updateNodeInternals(id);
+    }
+  }, [JSON.stringify(data.connectionHandles)]);
```
Comment (pcdavid) [inline packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/node/ImageNode.tsx]: Not sure about this, this looks like a potential performance issue.
Same remark for the other kinds of nodes.

**Review by pcdavid — CHANGES_REQUESTED** [review]

File: packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/node/ImageNode.tsx (hunk 22 above)
Comment (sbegaudeau) [reply packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/node/ImageNode.tsx]: - This code should not be there, we don't want to repeat that code in every single node, this belongs in a dedicated handle hook. Custom shapes will have to reuse it for example.
- Why do you need this ref? Why not a state?

I agree with @pcdavid, on top of being a potential performance issue, you do not have any idea of what would this behavior be in the future. That being said, it works for a large number of use cases it can be a pretty code solutions BUT, and I say this "but" with every single letter in upper case, I think that the unique place where this code would be needs a link to the original explanation which details why we need to be VERY VERY careful with this kind of solutions: https://github.com/facebook/react/issues/14476#issuecomment-471199055

Honestly, given the use case, I would rather have something like this:


```
const connectionHandlesIdentity = data.connectionHandles
                                      .map(handle => `${handle.nodeId}#${handle.edgeId}`)
                                      .join(', ');

useEffect(() => { ... }, [connectionHandlesIdentity])
```

The main difference being that if we change the structure of ConnectionHandle, we can select whether or not this code should be impacted.

Comment (mcharfadi) [reply packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/node/ImageNode.tsx]: Thanks for the explanation, I'll use a string as dependency of the useffect and put it in a custom hook then.

I used a ref because a state change would trigger a re render (unlike a ref change) and call updateNodeInternals which if called too much can impact the performance a lot. From my testing in this particular use case I get more render by using a state than a ref.

**Review by pcdavid — APPROVED** [review]
There are some known issues/possible improvements, but the basics are in place and work. @sbegaudeau unless you see something that jusitfies a veto from you, I say we merge this and continue improving in further PRs.

**Review by pcdavid — CHANGES_REQUESTED** [review]
Just to prevent this being merged before we decide how to proceed.
